src/DefinitionResolver.php

```diff
@@ -479,8 +483,10 @@ public function resolveExpressionNodeToType(Node\Expr $expr): Type
                 if ($expr instanceof Node\Expr\MethodCall) {
                     $fqn .= '()';
                 }
-                $def = $this->index->getDefinition($fqn);
-                if ($def !== null) {
+                if ($def = $this->index->getDefinition($fqn)) {
+                    if ($def->type instanceof Types\This || $def->type instanceof Types\Self_) {
+                        return $this->resolveExpressionNodeToType($expr->var);
+                    }
```

Comment:
Oh, sorry, I missed the `$this` keyword (I will have to rewrite a whole lot of stuff now :)) This is your call then - should I remove the check for `Self_`? Will there be any gain in removing it? I see only the downside of not supporting `return self` which even incorrectly is used a lot for indicating returnig `$this`. As I said - your call? 
As for `this` - I believe it is better this way - no point in breaking something that works. As said in the discussion earlier - some stuff (your example) is still not supported - so support is incomplete, but it will be a good foundation.